### PR TITLE
Issue 114 part 3 - child model simulation

### PR DESCRIPTION
**Pull request by M-Kusumgar**
have provided points of interests

### CODE REVIEW

File: inst/include/child_model_simulation.hpp [hunk 1]
```diff
@@ -392,46 +392,55 @@ void run_calculate_transmission_from_incidence_during_gestation(int t,
...
+  } // end a
 
-  if (cpars.mat_prev_input(t)) {
-    for (int a = 0; a < pars.base.options.p_fertility_age_groups; ++a) {
-      intermediate.children.age_weighted_hivneg += demog.age_specific_fertility_rate(a, t) / intermediate.children.asfr_sum * cpars.adult_female_hivnpop(a,t) ; //HIV negative 15-49 women weighted for ASFR
-      intermediate.children.age_weighted_infections +=  demog.age_specific_fertility_rate(a, t) / intermediate.children.asfr_sum  * cpars.adult_female_infections(a,t); //newly infected 15-49 women, weighted for ASFR
-    }//end
+  if (p_hc.mat_prev_input(t)) {
+    for (int a = 0; a < p_op.p_fertility_age_groups; ++a) {
+      auto asfr_weight = p_dm.age_specific_fertility_rate(a, t) / i_hc.asfr_sum;
```
Comment (M-Kusumgar) [inline inst/include/child_model_simulation.hpp:411]: new variable introduced

File: inst/include/child_model_simulation.hpp [hunk 2]
```diff
@@ -440,59 +449,55 @@ void run_calculate_perinatal_transmission_rate(int t,
...
+                         i_hc.PMTCT_coverage(6);
+
+  i_hc.no_PMTCT = 1 - i_hc.receiving_PMTCT;
+  i_hc.no_PMTCT = std::max(i_hc.no_PMTCT, 0.0);
+
+  // Transmission among women not on treatment
+  if (i_hc.num_wlhiv > 0) {
+    auto vertical_transmission = i_hc.prop_wlhiv_lt200 * p_hc.vertical_transmission_rate(4, 0) +
+                                 i_hc.prop_wlhiv_200to350 * p_hc.vertical_transmission_rate(2, 0) +
+                                 i_hc.prop_wlhiv_gte350 * p_hc.vertical_transmission_rate(0, 0);
```
Comment (M-Kusumgar) [inline inst/include/child_model_simulation.hpp:490]: new variable introduced

File: inst/include/child_model_simulation.hpp [hunk 3]
```diff
@@ -440,59 +449,55 @@ void run_calculate_perinatal_transmission_rate(int t,
...
+    i_hc.perinatal_transmission_rate += i_hc.no_PMTCT * vertical_transmission;
   }
```
Comment (M-Kusumgar) [inline inst/include/child_model_simulation.hpp:492]: dont need this else branch, it just does `x = x` same thing below

File: inst/include/child_model_simulation.hpp [hunk 4]
```diff
@@ -520,124 +526,116 @@ void run_bf_transmission_rate(int t,
...
+    i_hc.percent_no_treatment = std::max(i_hc.percent_no_treatment, 0.0);
+
+    // No treatment
+    if (p_hc.breastfeeding_duration_no_art(bf, t) < 1) {
+      auto vertical_transmission = i_hc.prop_wlhiv_lt200 * p_hc.vertical_transmission_rate(4, 1) +
+                                   i_hc.prop_wlhiv_200to350 * p_hc.vertical_transmission_rate(2,1) +
+                                   i_hc.prop_wlhiv_gte350 * p_hc.vertical_transmission_rate(0,1);
+      i_hc.bf_transmission_rate(index) += i_hc.percent_no_treatment *
+                                          vertical_transmission *
+                                          2 * (1 - p_hc.breastfeeding_duration_no_art(bf, t));
```
Comment (M-Kusumgar) [inline inst/include/child_model_simulation.hpp:632]: pulled out a factor of 2 here and defined the rest in the brackets as `vertical_transmission`

File: inst/include/child_model_simulation.hpp [hunk 5]
```diff
@@ -520,124 +526,116 @@ void run_bf_transmission_rate(int t,
...
     }
 
     if (bf < 1) {
-      intermediate.children.bf_transmission_rate(index) = intermediate.children.bf_transmission_rate(index)/ 4;
+      i_hc.bf_transmission_rate(index) /= 4;
```
Comment (M-Kusumgar) [inline inst/include/child_model_simulation.hpp:636]: using `/=` here

File: inst/include/child_model_simulation.hpp [hunk 6]
```diff
@@ -674,88 +674,99 @@ void run_child_hiv_infections(int t,
...
         }
-      }// end hc1DS
-      state_next.base.p_hiv_pop(0, s) +=  state_next.children.hiv_births * intermediate.children.perinatal_transmission_rate * demog.births_sex_prop(s,t);
-      state_next.base.p_infections(0, s) += state_next.children.hiv_births * intermediate.children.perinatal_transmission_rate * demog.births_sex_prop(s,t);
-    }// end NS
-
-    //Breastfeeding transmission
-    //0-6
+      } // end hc1DS
+      auto hiv_births_by_sex = perinatal_transmission_births * p_dm.births_sex_prop(s, t);
```
Comment (M-Kusumgar) [inline inst/include/child_model_simulation.hpp:700]: new variable defined

File: inst/include/child_model_simulation.hpp [hunk 7]
```diff
@@ -674,88 +674,99 @@ void run_child_hiv_infections(int t,
...
+        n_hc.hc1_hiv_pop(hd, 3, 2, s) += new_hc1_infections_24_plus;
+        n_ba.p_infections(2, s) += new_hc1_infections_24_plus;
+      } // end hc1DS
+
+      // 12-24
+      n_ba.p_hiv_pop(1, s) += bf_hiv_transmission_12_24;
+
+      // 24 plus
+      n_ba.p_hiv_pop(2, s) += bf_hiv_transmission_24_plus;
+    } // end NS
```
Comment (M-Kusumgar) [inline inst/include/child_model_simulation.hpp:768]: this entire thing is worth checking in your editor, github diff makes it look really messy, i have introduced several variables and factorised/simplified a decent amount of algebra

File: inst/include/child_model_simulation.hpp (hunk 2 above)
Comment (jeffeaton) [reply inst/include/child_model_simulation.hpp:490]: Suggest a different name for this for clarity. Something like `untreated_vert_transm_rate`. @mwalte10 also weigh in

File: inst/include/child_model_simulation.hpp (hunk 5 above)
Comment (jeffeaton) [reply inst/include/child_model_simulation.hpp:636]: Suggestion `4.0`

File: inst/include/child_model_simulation.hpp [hunk 8]
```diff
@@ -674,88 +674,99 @@ void run_child_hiv_infections(int t,
...
-      state_next.base.p_hiv_pop(1, s) += state_next.children.hiv_births * intermediate.children.bf_transmission_rate(2) *
-        (state_next.base.p_total_pop(1,s) - state_next.base.p_hiv_pop(1,s)) / ((state_next.base.p_total_pop(1,0) + state_next.base.p_total_pop(1,1)) - (state_next.base.p_hiv_pop(1,0) + state_next.base.p_hiv_pop(1,1)));
-      //24 plus
-      state_next.base.p_hiv_pop(2, s) +=  state_next.children.hiv_births  * (intermediate.children.bf_transmission_rate(3))*
-        (state_next.base.p_total_pop(2,s) - state_next.base.p_hiv_pop(2,s)) / ((state_next.base.p_total_pop(2,0) + state_next.base.p_total_pop(2,1)) - (state_next.base.p_hiv_pop(2,0) + state_next.base.p_hiv_pop(2,1)));
-    }// end NS
+    internal::run_bf_transmission_rate(t, pars, state_curr, state_next, intermediate, 12, ss_c.hBF, 3);
+    for (int s = 0; s < ss_b.NS; ++s) {
+      auto uninfected_prop_12_24 = (n_ba.p_total_pop(1, s) - n_ba.p_hiv_pop(1, s)) /
+                                   (n_ba.p_total_pop(1, 0) - n_ba.p_hiv_pop(1, 0) + n_ba.p_total_pop(1, 1) - n_ba.p_hiv_pop(1, 1));
```
Comment (jeffeaton) [inline inst/include/child_model_simulation.hpp:747]: This denominator could be defined once before the `for (int s = 0...`) (and same in line 749)

File: inst/include/child_model_simulation.hpp (hunk 7 above)
Comment (jeffeaton) [reply inst/include/child_model_simulation.hpp:768]: @mwalte10 -- do we have a test that would fail if something went awry in this refactor?

**Review by jeffeaton — COMMENTED** [review]
Thanks, looks good to me. But @mwalte10 will be a better critical reviewer on some of the questions raised.

File: inst/include/child_model_simulation.hpp (hunk 8 above)
Comment (M-Kusumgar) [reply inst/include/child_model_simulation.hpp:747]: sooooo im glad i wasnt the only one, i did this initially and that "Infections among children align" test in the child model tests failed and i was like whatttt, it took me an hour to find that this exact suggestion is what broke it haha, if you see at the bottom of this `for (int s = 0...)`, `p_hiv_pop` changes in that loop so its actually a variable quantity!! that took me way too long to figure out

File: inst/include/child_model_simulation.hpp (hunk 7 above)
Comment (M-Kusumgar) [reply inst/include/child_model_simulation.hpp:768]: something did go wrong in this refactor, the "Infections among children align" test in child model failed and i managed to fix it, but also some piece of mind, with all these tests, i run the child model on main, save that output object, then i save out the object from this branch with the child model running and i compare the two objects

the output is the exact same with a tolerance of 1e-10 (due to floating point errors)

File: inst/include/child_model_simulation.hpp (hunk 5 above)
Comment (M-Kusumgar) [reply inst/include/child_model_simulation.hpp]: yep done :)

**Review by r-ash — APPROVED** [review]
This one looks great too

File: inst/include/child_model_simulation.hpp (hunk 2 above)
Comment (mwalte10) [reply inst/include/child_model_simulation.hpp:490]: lets do untreated_vertical_tr?

File: inst/include/child_model_simulation.hpp (hunk 6 above)
Comment (mwalte10) [reply inst/include/child_model_simulation.hpp:700]: perinatal_births_by_sex

File: inst/include/child_model_simulation.hpp (hunk 2 above)
Comment (M-Kusumgar) [reply inst/include/child_model_simulation.hpp:490]: done

File: inst/include/child_model_simulation.hpp (hunk 1 above)
Comment (mwalte10) [reply inst/include/child_model_simulation.hpp:411]: good

**Review by mwalte10 — APPROVED** [review]
approved per call!

**Pull request merged**